File: nightlife_engine/core/content_generator.py
"""Main content generation engine."""

from typing import Dict, Any, Optional
from ..models.content_request import ContentRequest, ContentType
from ..models.brand_profile import BrandProfile
from ..models.event_context import EventContext
from .style_engine import StyleEngine
from .example_learner import ExampleLearner
from ..formatters import (
    InstagramReelFormatter,
    InstagramPostFormatter,
    PromotionFormatter,
    FlyerFormatter
)


class ContentGenerator:
    """
    Main content generation engine for nightlife social media.

    This is the primary interface for generating ready-to-post content.
    """

    def __init__(self):
        """Initialize content generator."""
        self.example_learner = ExampleLearner()

    def add_example(self, content: str, content_type: str) -> None:
        """
        Add example content to learn from.

        Args:
            content: The example content text
            content_type: Type of content (Instagram Reel, Instagram Post, etc.)
        """
        self.example_learner.add_example(content, content_type)

    def generate(self, request: ContentRequest) -> str:
        """
        Generate content based on the request.

        Args:
            request: ContentRequest with all necessary information

        Returns:
            Formatted, ready-to-post content string

        Raises:
            ValueError: If request is invalid or incomplete
        """
        # Initialize style engine
        style_engine = StyleEngine(request.brand_profile)

        # Get example guidance if available
        example_guidance = None
        if request.example_content:
            self.example_learner.add_example(
                request.example_content,
                request.content_type.value
            )
        example_guidance = self.example_learner.get_style_guidance(
            request.content_type.value
        )

        # Generate content based on type
        generated_content = self._generate_content_parts(
            request,
            style_engine,
            example_guidance
        )

        # Format output
        formatter = self._get_formatter(request)
        formatted_content = formatter.format(generated_content)

        return formatted_content

    def _generate_content_parts(
        self,
        request: ContentRequest,
        style_engine: StyleEngine,
        example_guidance: Optional[str]
    ) -> Dict[str, Any]:
        """
        Generate the individual parts of content.

        This method contains the core generation logic using the style guide
        and example patterns.
        """
        content_type = request.content_type

        if content_type == ContentType.INSTAGRAM_REEL:
            return self._generate_instagram_reel(request, style_engine, example_guidance)
        elif content_type == ContentType.INSTAGRAM_POST:
            return self._generate_instagram_post(request, style_engine, example_guidance)
        elif content_type == ContentType.PROMOTION:
            return self._generate_promotion(request, style_engine, example_guidance)
        elif content_type == ContentType.FLYER:
            return self._generate_flyer(request, style_engine, example_guidance)
        else:
            raise ValueError(f"Unsupported content type: {content_type}")

    def _generate_instagram_reel(
        self,
        request: ContentRequest,
        style_engine: StyleEngine,
        example_guidance: Optional[str]
    ) -> Dict[str, Any]:
        """Generate Instagram Reel content parts."""
        # This is a template-based generator
        # In a production system, this would integrate with an LLM
        # For now, we'll create structured templates

        event = request.event_context
        brand = request.brand_profile

        # Generate hook (max 8 words)
        hook = self._create_hook(event, brand, style_engine)

        # Generate main caption (1-2 sentences)
        caption = self._create_caption(event, brand, style_engine, "reel")

        # Generate CTA
        cta = self._create_cta(event, brand, style_engine)

        # Generate hashtags (exactly 5)
        hashtags = self._create_hashtags(brand, style_engine, 5)

        return {
            "hook": hook,
            "caption": caption,
            "cta": cta,
            "hashtags": hashtags
        }

    def _generate_instagram_post(
        self,
        request: ContentRequest,
        style_engine: StyleEngine,
        example_guidance: Optional[str]
    ) -> Dict[str, Any]:
        """Generate Instagram Post content parts."""
        event = request.event_context
        brand = request.brand_profile

        # Generate caption (max 2 sentences)
        caption = self._create_caption(event, brand, style_engine, "post")

        # Generate CTA
        cta = self._create_cta(event, brand, style_engine)

        # Generate hashtags (exactly 5)
        hashtags = self._create_hashtags(brand, style_engine, 5)

        return {
            "caption": caption,
            "cta": cta,
            "hashtags": hashtags
        }

    def _generate_promotion(
        self,
        request: ContentRequest,
        style_engine: StyleEngine,
        example_guidance: Optional[str]
    ) -> Dict[str, Any]:
        """Generate Promotion content parts."""
        event = request.event_context
        brand = request.brand_profile

        # Generate headline
        headline = self._create_headline(event, brand, style_engine)

        # Generate promo copy (2-3 sentences)
        promo_copy = self._create_promo_copy(event, brand, style_engine)

        # Generate CTA
        cta = self._create_cta(event, brand, style_engine)

        return {
            "headline": headline,
            "promo_copy": promo_copy,
            "cta": cta
        }

    def _generate_flyer(
        self,
        request: ContentRequest,
        style_engine: StyleEngine,
        example_guidance: Optional[str]
    ) -> Dict[str, Any]:
        """Generate Flyer content parts."""
        event = request.event_context
        brand = request.brand_profile

        # Generate headline
        headline = self._create_headline(event, brand, style_engine)

        # Generate subheadline
        subheadline = self._create_subheadline(event, brand, style_engine)

        # Generate event details
        event_details = self._create_event_details(event, brand)

        return {
            "headline": headline,
            "subheadline": subheadline,
            "event_details": event_details
        }

    def _create_hook(
        self,
        event: EventContext,
        brand: BrandProfile,
        style_engine: StyleEngine
    ) -> str:
        """Create a hook caption (max 8 words)."""
        # Template-based hook generation
        if event and event.event_name:
            return f"{event.event_name} is live"
        return "Tonight hits different"

    def _create_caption(
        self,
        event: EventContext,
        brand: BrandProfile,
        style_engine: StyleEngine,
        format_type: str
    ) -> str:
        """Create main caption."""
        sentences = []

        if event and event.dj_performer:
            sentences.append(f"{event.dj_performer} on the decks.")

        if event and event.special_notes:
            sentences.append(event.special_notes)
        else:
            # Default caption based on brand energy
            from ..models.brand_profile import BrandEnergy
            if brand.brand_energy == BrandEnergy.HIGH_ENERGY_PARTY:
                sentences.append("Energy unmatched.")
            elif brand.brand_energy == BrandEnergy.SEXY_MOODY:
                sentences.append("The vibe speaks for itself.")
            elif brand.brand_energy == BrandEnergy.CHILL_UPSCALE:
                sentences.append("Elevated experiences, curated for you.")
            else:
                sentences.append("Where culture meets nightlife.")

        # Limit to 2 sentences for most formats
        return " ".join(sentences[:2])

    def _create_cta(
        self,
        event: EventContext,
        brand: BrandProfile,
        style_engine: StyleEngine
    ) -> str:
        """Create call-to-action line."""
        from ..models.brand_profile import CTAStyle

        cta_map = {
            CTAStyle.RSVP: "RSVP via link in bio.",
            CTAStyle.BOOK: "Book your table now.",
            CTAStyle.PULL_UP: "Pull up tonight.",
            CTAStyle.TONIGHT: "See you tonight.",
            CTAStyle.SOFT_INVITE: "Link in bio."
        }

        return cta_map.get(brand.cta_style, "Link in bio.")

    def _create_hashtags(
        self,
        brand: BrandProfile,
        style_engine: StyleEngine,
        count: int
    ) -> list:
        """Create hashtags based on brand focus."""
        hashtags = []

        # Base nightlife hashtags
        base_tags = ["#nightlife", "#nightout", "#weekend", "#latenight", "#vibes"]

        # Focus-specific hashtags
        from ..models.brand_profile import ContentFocus
        focus_tags = {
            ContentFocus.EVENTS_DJS: ["#livemusic", "#dj", "#event", "#party"],
            ContentFocus.DRINKS_MENU: ["#cocktails", "#drinks", "#bar", "#mixology"],
            ContentFocus.CROWD_ATMOSPHERE: ["#atmosphere", "#vibes", "#crowd", "#experience"],
            ContentFocus.PROMOTIONS: ["#deals", "#promo", "#special", "#offer"]
        }

        # Additional filler tags for padding
        filler_tags = ["#music", "#party", "#nightclub", "#bar", "#drinks", "#djs", "#live"]

        # Add base tags
        hashtags.extend(base_tags[:3])

        # Add focus-specific tags
        for focus in brand.content_focus:
            if focus in focus_tags:
                hashtags.extend(focus_tags[focus][:2])

        # Pad with additional tags if needed
        if len(hashtags) < count:
            # Add remaining base tags
            for tag in base_tags[3:]:
                if len(hashtags) >= count:
                    break
                if tag not in hashtags:
                    hashtags.append(tag)

        # Add filler tags if still not enough
        if len(hashtags) < count:
            for tag in filler_tags:
                if len(hashtags) >= count:
                    break
                if tag not in hashtags:
                    hashtags.append(tag)

        # Return exactly the requested count
        return hashtags[:count]

    def _create_headline(
        self,
        event: EventContext,
        brand: BrandProfile,
        style_engine: StyleEngine
    ) -> str:
        """Create headline for promotion or flyer."""
        if event and event.event_name:
            return event.event_name.upper()
        return "TONIGHT"

    def _create_subheadline(
        self,
        event: EventContext,
        brand: BrandProfile,
        style_engine: StyleEngine
    ) -> str:
        """Create subheadline for flyer."""
        if event and event.dj_performer:
            return f"Featuring {event.dj_performer}"
        return "An Elevated Experience"

    def _create_promo_copy(
        self,
        event: EventContext,
        brand: BrandProfile,
        style_engine: StyleEngine
    ) -> str:
        """Create promotional copy (2-3 sentences)."""
        sentences = []

        if event and event.event_name:
            sentences.append(f"{event.event_name} returns.")

        if event and event.dj_performer:
            sentences.append(f"{event.dj_performer} sets the tone.")
        else:
            sentences.append("The city's best-kept secret.")

        sentences.append("Limited availability.")

        return " ".join(sentences[:3])

    def _create_event_details(
        self,
        event: EventContext,
        brand: BrandProfile
    ) -> str:
        """Create structured event details for flyer."""
        details = []

        if event:
            if event.event_name:
                details.append(f"EVENT: {event.event_name}")

            if event.date:
                details.append(f"DATE: {event.format_date()}")

            if event.dj_performer:
                details.append(f"FEATURING: {event.dj_performer}")

            if event.special_notes:
                details.append(f"DETAILS: {event.special_notes}")

        if not details:
            details.append("EVENT DETAILS")
            details.append("See link in bio for full information")

        return "\n".join(details)

    def _get_formatter(self, request: ContentRequest):
        """Get the appropriate formatter for the content type."""
        formatter_map = {
            ContentType.INSTAGRAM_REEL: InstagramReelFormatter,
            ContentType.INSTAGRAM_POST: InstagramPostFormatter,
            ContentType.PROMOTION: PromotionFormatter,
            ContentType.FLYER: FlyerFormatter
        }

        formatter_class = formatter_map.get(request.content_type)
        if not formatter_class:
            raise ValueError(f"No formatter for {request.content_type}")

        return formatter_class(
            request.brand_profile,
            request.event_context
        )

File: webapp/lib/core/content-generator.ts
/**
 * Main content generation engine
 */

import { ContentRequest, ContentType } from "../models/content-request";
import { BrandProfile, CTAStyle, ContentFocus } from "../models/brand-profile";
import { EventContext, formatDate } from "../models/event-context";
import { StyleEngine } from "./style-engine";
import { GeneratedContent } from "../formatters/base-formatter";
import { InstagramReelFormatter } from "../formatters/instagram-reel-formatter";
import { InstagramPostFormatter } from "../formatters/instagram-post-formatter";
import { PromotionFormatter } from "../formatters/promotion-formatter";
import { FlyerFormatter } from "../formatters/flyer-formatter";

export class ContentGenerator {
  generate(request: ContentRequest): string {
    const styleEngine = new StyleEngine(request.brandProfile);
    const generatedContent = this.generateContentParts(request, styleEngine);
    const formatter = this.getFormatter(request);
    return formatter.format(generatedContent);
  }

  private generateContentParts(
    request: ContentRequest,
    styleEngine: StyleEngine
  ): GeneratedContent {
    switch (request.contentType) {
      case ContentType.INSTAGRAM_REEL:
        return this.generateInstagramReel(request, styleEngine);
      case ContentType.INSTAGRAM_POST:
        return this.generateInstagramPost(request, styleEngine);
      case ContentType.PROMOTION:
        return this.generatePromotion(request, styleEngine);
      case ContentType.FLYER:
        return this.generateFlyer(request, styleEngine);
      default:
        throw new Error(`Unsupported content type: ${request.contentType}`);
    }
  }

  private generateInstagramReel(
    request: ContentRequest,
    styleEngine: StyleEngine
  ): GeneratedContent {
    const event = request.eventContext;
    const brand = request.brandProfile;

    return {
      hook: this.createHook(event, brand),
      caption: this.createCaption(event, brand, "reel"),
      cta: this.createCTA(event, brand),
      hashtags: this.createHashtags(brand, 5),
    };
  }

  private generateInstagramPost(
    request: ContentRequest,
    styleEngine: StyleEngine
  ): GeneratedContent {
    const event = request.eventContext;
    const brand = request.brandProfile;

    return {
      caption: this.createCaption(event, brand, "post"),
      cta: this.createCTA(event, brand),
      hashtags: this.createHashtags(brand, 5),
    };
  }

  private generatePromotion(
    request: ContentRequest,
    styleEngine: StyleEngine
  ): GeneratedContent {
    const event = request.eventContext;
    const brand = request.brandProfile;

    return {
      headline: this.createHeadline(event, brand),
      promoCopy: this.createPromoCopy(event, brand),
      cta: this.createCTA(event, brand),
    };
  }

  private generateFlyer(
    request: ContentRequest,
    styleEngine: StyleEngine
  ): GeneratedContent {
    const event = request.eventContext;
    const brand = request.brandProfile;

    return {
      headline: this.createHeadline(event, brand),
      subheadline: this.createSubheadline(event, brand),
      eventDetails: this.createEventDetails(event, brand),
    };
  }

  private createHook(event?: EventContext, brand?: BrandProfile): string {
    if (event?.eventName) {
      return `${event.eventName} is live`;
    }
    return "Tonight hits different";
  }

  private createCaption(
    event?: EventContext,
    brand?: BrandProfile,
    formatType: string = "post"
  ): string {
    const sentences: string[] = [];

    if (event?.djPerformer) {
      sentences.push(`${event.djPerformer} on the decks.`);
    }

    if (event?.specialNotes) {
      sentences.push(event.specialNotes);
    } else if (brand) {
      // Default caption based on brand energy
      const energyMap: Record<string, string> = {
        "high-energy party": "Energy unmatched.",
        "sexy & moody": "The vibe speaks for itself.",
        "chill & upscale": "Elevated experiences, curated for you.",
        "cultural & artistic": "Where culture meets nightlife.",
      };
      sentences.push(energyMap[brand.brandEnergy] || "Where culture meets nightlife.");
    }

    return sentences.slice(0, 2).join(" ");
  }

  private createCTA(event?: EventContext, brand?: BrandProfile): string {
    if (!brand) return "Link in bio.";

    const ctaMap: Record<CTAStyle, string> = {
      [CTAStyle.RSVP]: "RSVP via link in bio.",
      [CTAStyle.BOOK]: "Book your table now.",
      [CTAStyle.PULL_UP]: "Pull up tonight.",
      [CTAStyle.TONIGHT]: "See you tonight.",
      [CTAStyle.SOFT_INVITE]: "Link in bio.",
    };

    return ctaMap[brand.ctaStyle] || "Link in bio.";
  }

  private createHashtags(brand: BrandProfile, count: number): string[] {
    const hashtags: string[] = [];

    // Base nightlife hashtags
    const baseTags = [
      "#nightlife",
      "#nightout",
      "#weekend",
      "#latenight",
      "#vibes",
    ];

    // Focus-specific hashtags
    const focusTags: Record<ContentFocus, string[]> = {
      [ContentFocus.EVENTS_DJS]: ["#livemusic", "#dj", "#event", "#party"],
      [ContentFocus.DRINKS_MENU]: ["#cocktails", "#drinks", "#bar", "#mixology"],
      [ContentFocus.CROWD_ATMOSPHERE]: [
        "#atmosphere",
        "#vibes",
        "#crowd",
        "#experience",
      ],
      [ContentFocus.PROMOTIONS]: ["#deals", "#promo", "#special", "#offer"],
    };

    // Additional filler tags
    const fillerTags = [
      "#music",
      "#party",
      "#nightclub",
      "#bar",
      "#drinks",
      "#djs",
      "#live",
    ];

    // Add base tags
    hashtags.push(...baseTags.slice(0, 3));

    // Add focus-specific tags
    for (const focus of brand.contentFocus) {
      if (focusTags[focus]) {
        hashtags.push(...focusTags[focus].slice(0, 2));
      }
    }

    // Pad with remaining base tags if needed
    if (hashtags.length < count) {
      for (const tag of baseTags.slice(3)) {
        if (hashtags.length >= count) break;
        if (!hashtags.includes(tag)) {
          hashtags.push(tag);
        }
      }
    }

    // Add filler tags if still not enough
    if (hashtags.length < count) {
      for (const tag of fillerTags) {
        if (hashtags.length >= count) break;
        if (!hashtags.includes(tag)) {
          hashtags.push(tag);
        }
      }
    }

    return hashtags.slice(0, count);
  }

  private createHeadline(event?: EventContext, brand?: BrandProfile): string {
    if (event?.eventName) {
      return event.eventName.toUpperCase();
    }
    return "TONIGHT";
  }

  private createSubheadline(event?: EventContext, brand?: BrandProfile): string {
    if (event?.djPerformer) {
      return `Featuring ${event.djPerformer}`;
    }
    return "An Elevated Experience";
  }

  private createPromoCopy(event?: EventContext, brand?: BrandProfile): string {
    const sentences: string[] = [];

    if (event?.eventName) {
      sentences.push(`${event.eventName} returns.`);
    }

    if (event?.djPerformer) {
      sentences.push(`${event.djPerformer} sets the tone.`);
    } else {
      sentences.push("The city's best-kept secret.");
    }

    sentences.push("Limited availability.");

    return sentences.slice(0, 3).join(" ");
  }

  private createEventDetails(event?: EventContext, brand?: BrandProfile): string {
    const details: string[] = [];

    if (event) {
      if (event.eventName) {
        details.push(`EVENT: ${event.eventName}`);
      }

      if (event.date) {
        details.push(`DATE: ${formatDate(event.date)}`);
      }

      if (event.djPerformer) {
        details.push(`FEATURING: ${event.djPerformer}`);
      }

      if (event.specialNotes) {
        details.push(`DETAILS: ${event.specialNotes}`);
      }
    }

    if (details.length === 0) {
      details.push("EVENT DETAILS");
      details.push("See link in bio for full information");
    }

    return details.join("\n");
  }

  private getFormatter(request: ContentRequest) {
    const { brandProfile, eventContext } = request;

    switch (request.contentType) {
      case ContentType.INSTAGRAM_REEL:
        return new InstagramReelFormatter(brandProfile, eventContext);
      case ContentType.INSTAGRAM_POST:
        return new InstagramPostFormatter(brandProfile, eventContext);
      case ContentType.PROMOTION:
        return new PromotionFormatter(brandProfile, eventContext);
      case ContentType.FLYER:
        return new FlyerFormatter(brandProfile, eventContext);
      default:
        throw new Error(`No formatter for ${request.contentType}`);
    }
  }
}
